[RLlib] DreamerV3: docs (#37978)
ray-project/ray: rllib/algorithms/dreamerv3/README.md

# DreamerV3

## Overview
An RLlib-based implementation for TensorFlow/Keras of the "DreamerV3" model-based reinforcement

> learning algorithm by D. Hafner et al. (Google DeepMind) 2023

Reviewer: Link the arxiv paper

Author: done

> The implementation is based on RLlib's new Learner- and RLModule APIs and therefore allows
> for utilizing multi-GPU machines for training (see below for tips and tricks, example configs,
> and command lines).

Reviewer: ```suggestion
This implementation allows scaling up the training by using multi-GPU machines for neural network updates (see below for tips and tricks, example configs,
and command lines).
```

Author: done

DreamerV3 trains a world model in supervised fashion using real environment
interactions. The world model's objective is to correctly predict all aspects

> of the transition dynamics of the RL environment, which includes (besides predicting the
> correct next observations) predicting the received rewards as well as a boolean episode
> continuation flag.  

Reviewer: ```suggestion
of the transition dynamics of the RL environment, which includes predicting the
correct next observations, predicting the received rewards, and predicting a boolean for episode continuation.  
```

Author: fixed

> It utilizes a recurrent GRU-based architecture called "recurrent state space model" or RSSM.
> Alternatingly to training this world model, it is used to generate (or "dream") ficticious
> trajectories, with which two further models are trained in an RL-fashion:
> The critic and actor networks.

Reviewer: ```suggestion

For more specific details about the algorithm refer to the original paper. 
```

Author: fixed and added actual links to README.

Just like in a standard policy gradient algorithm (e.g. REINFORCE), the critic tries to
predict a correct value function (based on the world model-predicted rewards), whereas
the actor tries to come up with good actions to take for maximizing accumulated rewards
over time.
In other words, the actual RL components of the model (actor and critic) are never
trained on real environment data, but on dreamed trajectories only.

## Note on Hyperparameter Tuning for DreamerV3
DreamerV3 is an extremely versatile and stable algorithm that not only works well on

> different action- and observation spaces, including discrete and continuous actions, as well

Reviewer: ```suggestion
different action and observation spaces (i.e. discrete and continuous actions, as well
```

Author: done

> as image and vector observations, but also almost needs no hyperparameter tuning, except for

Reviewer: ```suggestion
as image and vector observations), but also has very little hyperparameters for  tuning.
```

Author: done

> a simple "model size" setting (from "XS" to "XL") and a value for the training ratio, which
> specifies how many steps to replay from the buffer for a training update vs how many
> steps to take in the actual environment.

Reviewer: ```suggestion
All you need is a simple "model size" setting (from "XS" to "XL") and a value for the training ratio, which
specifies how many steps to replay from the buffer for a training update vs how many
steps to take in the actual environment.
```

Author: done

## Note on multi-GPU Training with DreamerV3
We found that when using multiple GPUs for DreamerV3 training, the following simple
adjustments should be made to the standard config.

> - Multiply the original batch size (B=16) by the number of GPUs you are using.

Author: done

- Multiply the number of environments you sample from in parallel by the number of GPUs you are using.

> - Use a learning rate schedule for all learning rates (world model, actor, critic) with "priming".

Reviewer: Can you link `priming` to the relevant arxiv link for the papers here?

Author: done

  - In particular, the first 10k timesteps should use low rates of `0.4` times of the published rates
    (i.e. world model: `4e-5`, critic and actor: `1.2e-5`). 
  - Over the course of the next 10k timesteps, linearly increase all rates to
    n times their published values, where `n=max(4, [num GPUs])`.

## Example Configs and Command Lines
Use the config examples and templates in
[the tuned_examples folder here](https://github.com/ray-project/ray/tree/master/rllib/tuned_examples/dreamerv3)
in combination with the following scripts and command lines in order to run RLlib's DreamerV3 algorithm in your experiments:

### Atari100k
```shell
$ cd ray/rllib/tests

> $ python run_regression_tests.py --dir ../tuned_examples/dreamerv3/atari_100k.py --env ALE/Pong-v5 

Author: Yeah, good point. But we should add these WandB options to `rllib train` as well. These are super useful.

```

### DeepMind Control Suite (vision)
```shell
$ cd ray/rllib/tests
$ python run_regression_tests.py --dir ../tuned_examples/dreamerv3/dm_control_suite_vision.py --env DMC/cartpole/swingup 
```
Other `--env` options for the DM Control Suite would be `--env DMC/hopper/hop`, `--env DMC/walker/walk`, etc..
Note that you can also switch on WandB logging with the above script via the options
`--wandb-key=[your WandB API key] --wandb-project=[some project name] --wandb-run-name=[some run name]`

## Results
Our results on the Atari 100k and (visual) DeepMind Control Suite benchmarks match those
reported in the paper.

### Pong-v5 (100k) 1GPU vs 2GPUs vs 4GPUs

> <img src="https://github.com/ray-project/ray/tree/master/doc/source/rllib/images/dreamerv3/pong_1_2_and_4gpus.svg">

Reviewer: This is not the best practice as this image will always be pointing to master. So if we release 2.7 and tomorrow change the image, the old content will point to the new image on master. 

We should do sth like: 
![pong_1_2_and_4gpus](relative path to image)

Author: I think this is difficult doing this from the README (which is in a completely different location than the docs and does not get built together with the docs), but let me check.

Author: Figured it out. Never mind.
